vnext/Microsoft.ReactNative/JsiApi.cpp

```diff
@@ -325,15 +328,34 @@ facebook::jsi::JSError const &jsError) {                             \
 /*static*/ ReactNative::JsiRuntime JsiRuntime::GetOrCreate(
     std::shared_ptr<facebook::jsi::RuntimeHolderLazyInit> const &jsiRuntimeHolder,
     std::shared_ptr<facebook::jsi::Runtime> const &jsiRuntime) noexcept {
-  std::scoped_lock lock{s_mutex};
-  auto it = s_jsiRuntimeMap.find(reinterpret_cast<uintptr_t>(jsiRuntime.get()));
-  if (it != s_jsiRuntimeMap.end()) {
-    return it->second.get();
-  } else {
-    ReactNative::JsiRuntime abiJsiResult{make<JsiRuntime>(Mso::Copy(jsiRuntimeHolder), Mso::Copy(jsiRuntime))};
-    s_jsiRuntimeMap.try_emplace(reinterpret_cast<uintptr_t>(jsiRuntime.get()), abiJsiResult);
-    return abiJsiResult;
+  {
+    std::scoped_lock lock{s_mutex};
+    auto it = s_jsiRuntimeMap.find(reinterpret_cast<uintptr_t>(jsiRuntime.get()));
+    if (it != s_jsiRuntimeMap.end()) {
+      return it->second.get();
+    }
   }
+
+  return Create(jsiRuntimeHolder, jsiRuntime);
+}
+
+/*static*/ ReactNative::JsiRuntime JsiRuntime::Create(
+    std::shared_ptr<facebook::jsi::RuntimeHolderLazyInit> const &jsiRuntimeHolder,
+    std::shared_ptr<facebook::jsi::Runtime> const &jsiRuntime) noexcept {
+  auto buffer = std::make_shared<facebook::jsi::StringBuffer>(R"JS(
+      var __jsi_pal = function() {
+        var hostFunctionId = Symbol('hostFunctionId');
+        return {
+          getHostFunctionId: function(func) { return func[hostFunctionId]; },
+          setHostFunctionId: function(func, id) { Object.defineProperty(func, hostFunctionId, { value: id }); }
+        };
+      }();
+    )JS");
+  jsiRuntime->evaluateJavaScript(buffer, "JSI_PAL");
```

Comment:
where is `JSI_PAL` coming from?

Reply:
We need some URL as a source for the JavaScript. I just used this string. It would be better to use a URI instead. It would be interesting to see if we can create an URI that references DLL resources instead.

---
In reply to: [537225214](https://github.com/microsoft/react-native-windows/pull/6693#discussion_r537225214) [](ancestors = 537225214)

Reply:
got it makes sense. Maybe add a comment to say that, or name this something like `"FromJSI_not_a_real_file"`?